### PR TITLE
FHB-736 : Referral API

### PR DESCRIPTION
**Pull request by Zac-AND-Digital**
Ticket: [FHB-736](https://dfedigital.atlassian.net/browse/FHB-736)

---

- Convert Moq -> NSubstitute for Referral API
- Refactor & Cleanup Code

[FHB-736]: https://dfedigital.atlassian.net/browse/FHB-736?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Top-Cat — APPROVED** [review]

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenCreateOrUpdateService.cs [hunk 1]
```diff
@@ -35,17 +26,14 @@ public async Task ThenCreateServiceWithExistingOrganisation()
         var serviceDto = Mapper.Map<ReferralServiceDto>(service);
 
         CreateOrUpdateServiceCommand createOrUpdateOrganisationCommand = new(serviceDto);
-        CreateOrUpdateServiceCommandHandler handler = new(TestDbContext, Mapper, new Mock<ILogger<CreateOrUpdateServiceCommandHandler>>().Object);
+        CreateOrUpdateServiceCommandHandler handler = new(TestDbContext, Mapper, Substitute.For<ILogger<CreateOrUpdateServiceCommandHandler>>());
 
         //Act
-        var result = await handler.Handle(createOrUpdateOrganisationCommand, new System.Threading.CancellationToken());
+        var result = await handler.Handle(createOrUpdateOrganisationCommand, new CancellationToken());
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenCreateOrUpdateService.cs:32]: Could use CancellationToken.None instead!?

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs [hunk 2]
```diff
@@ -17,7 +17,7 @@ public async Task ThenGetReferralByIdOnly()
         GetReferralByIdCommandHandler handler = new(TestDbContext, Mapper);
 
         //Act
-        var result = await handler.Handle(command, new System.Threading.CancellationToken());
+        var result = await handler.Handle(command, new CancellationToken());
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs:20]: Could use CancellationToken.None!?

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs [hunk 3]
```diff
@@ -42,16 +42,16 @@ public async Task ThenGetReferralByIdOnly()
...
         await CreateReferral();
         var referral = TestDataProvider.GetReferralDto();
 
-        GetReferralsByOrganisationIdCommand command = new(2, referralOrderBy, isAssending, false,  1,10);
+        GetReferralsByOrganisationIdCommand command = new(2, referralOrderBy, isAscending, false,  1,10);
         GetReferralsByOrganisationIdCommandHandler handler = new(TestDbContext, Mapper);
 
         //Act
-        var result = await handler.Handle(command, new System.Threading.CancellationToken());
+        var result = await handler.Handle(command, new CancellationToken());
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs:54]: Could use CancellationToken.None!? Same elsewhere

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs [hunk 4]
```diff
@@ -31,11 +31,11 @@ public async Task ThenCreateASingleUserAccount()
         result.Should().BeGreaterThan(0);
         var actualUserAccount = await TestDbContext.UserAccounts
             .Include(x => x.OrganisationUserAccounts)
-            .FirstAsync(x => x.OrganisationUserAccounts != null && x.OrganisationUserAccounts.Any(x => x.OrganisationId == userAccountDto.OrganisationUserAccounts[0].Organisation.Id));
+            .FirstAsync(x => x.OrganisationUserAccounts != null && x.OrganisationUserAccounts.Any(uAo => uAo.OrganisationId == userAccountDto.OrganisationUserAccounts[0].Organisation.Id));
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs:34]: Maybe oua? uAo doesn't make sense and looks odd

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs [hunk 5]
```diff
@@ -153,11 +152,11 @@ public async Task ThenCreateUserAccounts()
         result.Should().BeTrue();
         var actualUserAccount = await TestDbContext.UserAccounts
             .Include(x => x.OrganisationUserAccounts)
-            .FirstAsync(x => x.OrganisationUserAccounts != null && x.OrganisationUserAccounts.Any(x => x.OrganisationId == userAccountDto.OrganisationUserAccounts[0].Organisation.Id));
+            .FirstAsync(x => x.OrganisationUserAccounts != null && x.OrganisationUserAccounts.Any(uAo => uAo.OrganisationId == userAccountDto.OrganisationUserAccounts[0].Organisation.Id));
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs:155]: Same as mentioned before

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs [hunk 6]
```diff
@@ -230,11 +229,11 @@ public async Task ThenUpdateUserAccounts()
...
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs:232]: Same as mentioned before

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.UnitTests/BaseUnitTest.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
+using AutoMapper;
+using FamilyHubs.Referral.Core;
+using FamilyHubs.ReferralService.Shared.Dto;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+
+namespace FamilyHubs.Referral.UnitTests;
+
+public class BaseUnitTest<T>
```
Comment (stevesatdfe) [inline src/service/referral-api/tests/FamilyHubs.ReferralApi.UnitTests/BaseUnitTest.cs:9]: Should this be abstract?

**Review by stevesatdfe — CHANGES_REQUESTED** [review]

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs (hunk 2 above)
Comment (Zac-AND-Digital) [reply src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingGetReferral.cs]: Resolved: [7593257](https://github.com/DFE-Digital/fh-services/pull/154/commits/759325774662a684fa86516f6b76feb93e986315)

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs (hunk 6 above)
Comment (Zac-AND-Digital) [reply src/service/referral-api/tests/FamilyHubs.ReferralApi.Integration.Tests/WhenUsingUserAccounts.cs]: Resolved: [7593257](https://github.com/DFE-Digital/fh-services/pull/154/commits/759325774662a684fa86516f6b76feb93e986315)

File: src/service/referral-api/tests/FamilyHubs.ReferralApi.UnitTests/BaseUnitTest.cs (hunk 7 above)
Comment (Zac-AND-Digital) [reply src/service/referral-api/tests/FamilyHubs.ReferralApi.UnitTests/BaseUnitTest.cs]: Resolved: [7593257](https://github.com/DFE-Digital/fh-services/pull/154/commits/759325774662a684fa86516f6b76feb93e986315)

**Review by stevesatdfe — APPROVED** [review]

**Review by stevesatdfe — APPROVED** [review]

**Review by Top-Cat — APPROVED** [review]

**Pull request merged**